### ISSUE TITLE
Uncaught mismatch between Type[T] and T in function arguments

### ISSUE THREAD

**Issue by hcarty** [microsoft/pyright#1400]
**Describe the bug**
When a class and an object of that class type are specified as function argument type constraints, function calls do not seem to be properly type constraints.

**To Reproduce**
See example code below.  The last two lines will type check when they should not because the class and object arguments do not match.  Type checking for `make` behaves like I would expect but `op` does not - the type of the `c` parameter does not seem to affect constraints applied to the `o` parameter.

The `make` function type checks as I would expect, but in that case the class is a function parameter and the object is returned rather than being another function parameter.

**Expected behavior**
I would expect the last two lines, the final calls to `op`, in the example below to fail to type check.

**Screenshots or Code**
```python
# pyright: strict

from typing import Type, TypeVar

class A:
    a: int

class B:
    b: str

T = TypeVar("T")

def make(c: Type[T]) -> T:
    ...

S = TypeVar("S")

def op(c: Type[S], o: S) -> None:
    ...

# These type check as I would expect
a: A = make(A)
b: B = make(B)
op(A, a)
op(B, b)

# These fail to type check as I would expect
a_invalid: A = make(B)
b_invalid: B = make(A)

# These type check when I would expect them to fail
op(B, a)
op(A, b)
```

**VS Code extension or command-line**
Pylance in VS Code.  Pylance version 2021.1.1 which says it's using Pyright 1.1.101 plus a few extra changes.

**Additional context**
mypy 0.790 has the same behavior as Pyright currently does.

**Comment by erictraut** [comment]
Thanks for the bug report.

As counterintuitive as it might seem, I think pyright is doing the right thing in these cases. Its behavior matches that of mypy, another popular type checker.

The reason the last two lines in your sample type check is that the TypeVar constraint solver is able to find a type for `S` that satisfies all of the constraints. In this particular case, that type is `Union[A, B]`. (Mypy's constraint solver uses a slightly different technique but comes up with yet another solution that works: `object`.)

The reason this works is that the type variable for `Type` is considered covariant, which means the constraint solver is able to "widen" the type when solving the constraints. Mypy widens the type by "joining" the subtypes (i.e. finding the narrowest type that is an ancestor of both `A` and `B`, which turns out to be `object`). Pyright widens the type by creating a union of the two types. Both are legal solutions.

I started to explore potential workarounds for what you're trying to do, and I may have uncovered a different bug in pyright — or at least a case where its behavior differs from mypy. I was thinking that maybe you could use a `Callable[[], A]` type annotation rather than `Type[A]`. Something like this:

```python
def op2(c: Callable[[], S], o: S) -> S:
    ...

op2(B, a)
op2(A, b)
```

Mypy generates an error in this case, but pyright does not. I'm not sure why mypy treats this case differently. I thought that the return type of a callable was considered covariant, so I'd think that it would give the same results as `Type[S]`.

I'd like to pull in some other folks who have deep knowledge of the Python type system. Perhaps @gvanrossum or @JukkaL can offer some wisdom here.

**Comment by hcarty** [comment]
Thank you for giving such a detailed reply!  I had read that [TypeVar is invariant](https://docs.python.org/3/library/typing.html#typing.TypeVar) by default and mistakenly expected `Type[T]` to act similarly.  After reading the appropriate section of PEP-484 I see that `Type[T]`'s covariance is specified explicitly.

**Comment by gvanrossum** [comment]
I'm not sure that we can trust mypy sufficiently in this case; I have seen bugs there when it comes to classes vs. functions with a similar signature. I really think we'd need @JukkaL's input but that may be hard (he's very focused on other things ATM).

FWIW the variance of `T` is not relevant to the question of whether `Type[T]` is variant -- if `Type[]` was specified as a regular generic class (which it isn't), its variance would be decided by that of the type variable used in its class definition. Fortunately you found the relevant text about this in PEP 484.

**Comment by erictraut** [comment]
Thanks Guido. I'm pretty sure that Pyright is doing the right thing here. I'm going to mark this issue as "as designed" and close it. If Jukka provides additional information to the contrary, I'm happy to revisit it.
